### PR TITLE
Fix: remove NaN points before passing to points layer

### LINKED ISSUE #446: Unexpected "tail" from previous time frame showing in points layer in multi-animal datasets
Steps:
1- `napari -w movement`
2- Load the sample data file `SLEAP_two-mice_octagon.analysis.h5`
3- The frame slider is at frame 256, but I can see keypoints for both frame 255 and frame 256. Notice there are more keypoints per individual than expected (the corresponding `movement` dataset says 7 keypoints)

kpt_256
kpt_255

I also get this error message:
Screenshot 2025-02-25 at 17 01 39

The issue is sometimes fixed when loading the video. It is so far always fixed if I load the video and then flick back and forth a few frames.

Note that in the screenshots above, the total number of frames does not correspond to the video length - seems like an issue with the order of the axes when the video is not loaded?

_Originally posted by @sfmig in https://github.com/neuroinformatics-unit/movement/issues/393#issuecomment-2682694542_

----

As far as I can tell, this only happens with SLEAP multi-animal datasets. Interestingly, it happens with the "mixed-labels" version of the Aeon dataset, but not with the "proofread" one. I wonder whether this is somehow related to mixed identities/tracks in SLEAP files. Intriguingly, whether we are using the analysis.h5 or the .slp file doesn't matter. 

_Originally posted by @niksirbi in https://github.com/neuroinformatics-unit/movement/issues/393#issuecomment-2682897276_

----

To me it **happens with all the multi-animal datasets**, not only SLEAP ones (the Sherlock bboxes dataset - which is not in GIN but can share- , and the DLC two mice dataset). I think maybe that is a stronger hint.

Another hint is that the frame slider end is set to 512 in all cases - seems like a default? It makes me think napari is interpreting the axes wrong... Just to be extra clear I am not loading any videos, just data.

_Originally posted by @sfmig in https://github.com/neuroinformatics-unit/movement/issues/393#issuecomment-2684407905_

----

Yay, you make a strong case. If it's all multi-animal datasets, and the problem is solved by pre-loading a video first, it is indeed highly likely that napari does something funky with the axes there. I guess if the video is added first and the points later, napari aligns the axes of the points layer to the image layer? Do you know if pre-loading the image only (still frame) also solves this or do we always need the video?

_Originally posted by @niksirbi in https://github.com/neuroinformatics-unit/movement/issues/393#issuecomment-2684499364_
            
----

I just tried with the two mice octagon data - the problem still exists if you load a still frame. **The number of frames in the slider does not match the number of frames in the data.**

_Originally posted by @sfmig in https://github.com/neuroinformatics-unit/movement/issues/393#issuecomment-2687857507_

### PR DESCRIPTION
**Pull request by sfmig**
## Description

**What is this PR**

- [x] Bug fix
- [ ] Addition of a new feature
- [ ] Other

**Why is this PR needed?**
- We noticed the dimension slider was off when loading data with NaNs, and points appeared "duplicated" per frame
- This was the case for our multi-animal datasets.
- A workaround is to exclude the data points with nans before creating the points layer.

**What does this PR do?**
- Remove nans before passing the data to points layers, and ensure the frame slider matches the full range of data.
    - a frame should be visible in the slider even if there is all data is nan
- Adds a test to ensure the dimensions slider matches the n frames
    - ~This is slightly us writing tests for code we didn't write... let me know thoughts~ --> not anymore, since now we enforce the frame slider to match the full range of frames
- Suggest new names for some of the napari modules (removing `_` and shorter) 
    - Happy to hear thoughts on this too.

## References

Fixes #446 

## How has this PR been tested?

Tests pass locally and in CI.

## Is this a breaking change?

No.

## Does this PR require an update to the documentation?

No.

## Checklist:

- [x] The code has been tested locally
- [ ] Tests have been added to cover all new functionality
- [x] The documentation has been updated to reflect any changes
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

### CODE REVIEW

File: tests/test_unit/test_napari_plugin/test_poses_loader_widget.py [hunk 1]
```diff
@@ -167,3 +168,28 @@ def test_on_load_clicked_with_valid_file_path(
     # Check that a Points layer was added to the viewer
     points_layer = poses_loader_widget.viewer.layers[0]
     assert points_layer.name == f"poses: {file_path.name}"
+
+
+def test_dimension_slider_matches_frames(make_napari_viewer_proxy):
```
Comment (niksirbi) [inline tests/test_unit/test_napari_plugin/test_poses_loader_widget.py:173]: Having this test is good for my sanity, let's keep it in, even if it tests "someone else's code".

File: tests/test_unit/test_napari_plugin/test_poses_loader_widget.py [hunk 2]
```diff
@@ -167,3 +168,28 @@ def test_on_load_clicked_with_valid_file_path(
...
+    ds = load_poses.from_file(file_path, "SLEAP", fps=1)
+    assert ds.position.isnull().any()
+
+    # Call the _on_load_clicked method (pretend the user clicked "Load")
+    poses_loader_widget._on_load_clicked()
+
+    # Check the frame slider is set to the correct value
+    assert viewer.dims.range[0] == RangeTuple(
+        start=0.0, stop=ds.position.shape[0] - 1, step=1.0
+    )
```
Comment (niksirbi) [inline tests/test_unit/test_napari_plugin/test_poses_loader_widget.py:195]: Just a clarification here:

What would happen if say the very first or the very last frame contained only NaNs? My understanding of the code tells me that the frame would be completely dropped, causing the min-max range of frames to change, making this assertion to fail.

It's not the case here because this edge case doesn't manifest in the dataset you chose for this test.

Perhaps we should think of a more comprehensive test to check for this, but we shouldn't do it here in this PR, because I'm keen to get this bugfix merged soon.

File: movement/napari/_meta_widget.py [hunk 3]
```diff
@@ -3,7 +3,7 @@
 from napari.viewer import Viewer
 from qt_niu.collapsible_widget import CollapsibleWidgetContainer
 
-from movement.napari._loader_widgets import PosesLoader
+from movement.napari.widgets import PosesLoader
```
Comment (niksirbi) [inline movement/napari/_meta_widget.py:6]: I will use this code line to comment on the renaming of the module.

I have several points to make:
- I originally went with `_loader_widgets` because I expect us to have many widgets for different jobs (filtering, defining ROIs, saving data, etc.) and I wouldn't want them all to end up in a giant file named "widgets". But perhaps this is me anticipating too much.
- My main motivation for having the widget-related module private was to easily exclude them from the API docs. If you render the docs in your branch, you will see that the `PosesLoader` class is documented with hundreds of methods and attributes (inherited from qt), and I don't think that's useful. Perhaps you can remove the underscore from the file name, but instead explicitly exclude this file from the API docs (there is a `exlude_module` list in the `make_api_index.py` script).
- Whatever you decide to do with with this module, we should probably also do for the meta-widget module. For example, if we have `loader_widgets.py`, we should also have `meta_widget.py` (and both of them should be excluded from the API docs).

**Review by niksirbi — APPROVED** [review]
Thanks @sfmig for the speedy and necessary bug fix! And many thanks to @willGraham01 as well who helped with the debugging.

I think this PR is basically good to go, and I'd rather merge this soon-ish (given how annoying the bug is). I have added specific comments with my thoughts on the new test and the renaming of the modules.

File: movement/napari/_meta_widget.py (hunk 3 above)
Comment (sfmig) [reply movement/napari/_meta_widget.py:6]: yes, I agree excluding from the API and renaming both previously private modules is the best option.

File: tests/test_unit/test_napari_plugin/test_poses_loader_widget.py (hunk 2 above)
Comment (sfmig) [reply tests/test_unit/test_napari_plugin/test_poses_loader_widget.py:195]: oh yes good point, can't believe I missed that!

I'd actually say let's not rush it and comprehensively test this in this PR

**Comment by sfmig** [comment]
thanks @niksirbi - just tagging you for a quick second review since I added:
- an extended test to check the cases I had missed
- a few lines to ensure the frame slider always represents the full range of frames available in the movement dataset

**Review by niksirbi — APPROVED** [review]
Great job @sfmig, it's now much better implemented and painstakingly tested. I trust this fix now.
I also put it through its paces by loading every conceivable dataset, and it works every time!
A good test for me was loading the DLC vs SLEAP version of the EPM dataset (because the initial frames are NaN in the SLEAP version but not in the DLC version). Glad to report that the slider range is the same in both cases, and all is as expected.

This is good to merge now 🎉
